GoogleUtilities/SceneDelegateSwizzler/GULSceneDelegateSwizzler.m

```diff
@@ -0,0 +1,466 @@
+typedef void (^GULSceneDelegateInterceptorCallback)(id<UISceneDelegate>);
+
+// The strings below are the keys for associated objects.
+static char const *const kGULRealIMPBySelectorKey = "GUL_realIMPBySelector";
+static char const *const kGULRealClassKey = "GUL_realClass";
+
+static GULLoggerService kGULLoggerSwizzler = @"[GoogleUtilities/SceneDelegateSwizzler]";
+
+// Since Firebase SDKs also use this for app delegate proxying, in order to not be a breaking change
+// we disable App Delegate proxying when either of these two flags are set to NO.
+
+/** Plist key that allows Firebase developers to disable App and Scene Delegate Proxying. */
+static NSString *const kGULFirebaseSceneDelegateProxyEnabledPlistKey =
+    @"FirebaseAppDelegateProxyEnabled";
+
+/** Plist key that allows developers not using Firebase to disable App and Scene Delegate Proxying.
+ */
+static NSString *const kGULGoogleUtilitiesSceneDelegateProxyEnabledPlistKey =
+    @"GoogleUtilitiesAppDelegateProxyEnabled";
+
+/** The prefix of the Scene Delegate. */
+static NSString *const kGULSceneDelegatePrefix = @"GUL_";
+
+/**
+ * This class is necessary to store the delegates in an NSArray without retaining them.
+ * [NSValue valueWithNonRetainedObject] also provides this functionality, but does not provide a
+ * zeroing pointer. This will cause EXC_BAD_ACCESS when trying to access the object after it is
+ * dealloced. Instead, this container stores a weak, zeroing reference to the object, which
+ * automatically is set to nil by the runtime when the object is dealloced.
+ */
+@interface GULSceneZeroingWeakContainer : NSObject
+
+/** Stores a weak object. */
+@property(nonatomic, weak) id object;
+
+@end
+
+@implementation GULSceneZeroingWeakContainer
+@end
+
+@implementation GULSceneDelegateSwizzler
+
+#pragma mark - Public methods
+
++ (BOOL)isSceneDelegateProxyEnabled {
+  NSDictionary *infoDictionary = [NSBundle mainBundle].infoDictionary;
+
+  id isFirebaseProxyEnabledPlistValue =
+      infoDictionary[kGULFirebaseSceneDelegateProxyEnabledPlistKey];
+  id isGoogleProxyEnabledPlistValue =
+      infoDictionary[kGULGoogleUtilitiesSceneDelegateProxyEnabledPlistKey];
+
+  // Enabled by default.
+  BOOL isFirebaseSceneDelegateProxyEnabled = YES;
+  BOOL isGoogleUtilitiesSceneDelegateProxyEnabled = YES;
+
+  if ([isFirebaseProxyEnabledPlistValue isKindOfClass:[NSNumber class]]) {
+    isFirebaseSceneDelegateProxyEnabled = [isFirebaseProxyEnabledPlistValue boolValue];
+  }
+
+  if ([isGoogleProxyEnabledPlistValue isKindOfClass:[NSNumber class]]) {
+    isGoogleUtilitiesSceneDelegateProxyEnabled = [isGoogleProxyEnabledPlistValue boolValue];
+  }
+
+  // Only deactivate the proxy if it is explicitly disabled by app developers using either one of
+  // the plist flags.
+  return isFirebaseSceneDelegateProxyEnabled && isGoogleUtilitiesSceneDelegateProxyEnabled;
+}
+
++ (GULSceneDelegateInterceptorID)registerSceneDelegateInterceptor:(id<UISceneDelegate>)interceptor {
```

Comment:
In the future we need to think how to avoid code duplication with `GULAppDelegateSwizzler`

Reply:
My general idea is to have a single class that takes care of Application delegate, Scene delegate and even more delegates in the future, since there are many shared logic.
While for now, to solve the scene delegate issue and at the same time minimize the impact on App delegate swizzler, we just created a standalone class for scene delegate.
It depends to duplicate the code or not. Please see my comments in each case above.